xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h

```diff
@@ -294,6 +294,94 @@ class CSettingValue
 namespace kodi {
 namespace addon {
 
+/*
+ * Internally used helper class to manage processing of a "C" structure in "CPP"
+ * class.
+ *
+ * At constant, the "C" structure is copied, otherwise the given pointer is
+ * superseded and is changeable.
+ *
+ * -----------------------------------------------------------------------------
+ *
+ * Example:
+ *
+ * ~~~~~~~~~~~~~{.cpp}
+ * extern "C" typedef struct C_SAMPLE_DATA
+ * {
+ *   unsigned int iUniqueId;
+ * } C_SAMPLE_DATA;
+ *
+ * class CPPSampleData : public CStructHdl<CPPSampleData, C_SAMPLE_DATA>
+ * {
+ * public:
+ *   CPPSampleData() = default;
+ *   CPPSampleData(const CPPSampleData& sample) : CStructHdl(sample) { }
+ *   CPPSampleData(const C_SAMPLE_DATA* sample) : CStructHdl(sample) { }
+ *   CPPSampleData(C_SAMPLE_DATA* sample) : CStructHdl(sample) { }
+ *
+ *   void SetUniqueId(unsigned int uniqueId) { m_cStructure->iUniqueId = uniqueId; }
+ *   unsigned int GetUniqueId() const { return m_cStructure->iUniqueId; }
+ * };
+ *
+ * ~~~~~~~~~~~~~
+ *
+ * The class automatically works with a "C" = "CPP" or "CPP" = "C".
+ */
+template <class CPP_CLASS, typename C_STRUCT>
+class CStructHdl
+{
+public:
+  CStructHdl()
+    : m_owner(true),
+      m_cStructure(new C_STRUCT)
+  {
+  }
+
+  CStructHdl(const CPP_CLASS& cppClass)
+    : m_owner(true),
+      m_cStructure(new C_STRUCT(*cppClass.m_cStructure))
+  {
+  }
+
+  CStructHdl(const C_STRUCT* cStructure)
+    : m_owner(true),
+      m_cStructure(new C_STRUCT({*cStructure}))
+  {
+  }
+
+  CStructHdl(C_STRUCT* cStructure)
+    : m_owner(false),
+      m_cStructure(cStructure)
+  {
+  }
+
+  const CPP_CLASS& operator =(const CPP_CLASS& right)
+  {
```

Comment:
```c++
if (m_cStructure && m_owner)
{
  delete m_cStructure;
  m_cStructure = nullptr;
}
```

before changing m_owner and assigning m_cStructure?

Reply:
Also here to set with nullptr bring hidden errors / crashes.

Only this `m_owner` in the class can really make private instead of protected.

Reply:
Ah you mean? Yes that could be more safe to prevent memleaks.
```cpp
  const CPP_CLASS& operator =(const CPP_CLASS& right)
  {
    if (m_cStructure && m_owner)
      delete m_cStructure;
    m_owner = true;
    m_cStructure = new C_STRUCT(*right.m_cStructure);
    return *this;
  }
```

Reply:
The check in the if condition for `m_cStructure` is also not necessary. Calling `delete` with a `nullptr` is well defined since C++14:

> If expression evaluates to a null pointer value, no destructors are called, and the deallocation function is not called.  
> https://en.cppreference.com/w/cpp/language/delete

Reply:
Right, but why should I call delete if I know it will do nothing?

Reply:
> Calling delete with a nullptr is well defined since C++14:

 Kodi is c++ 11

Reply:
No more since yesterday https://github.com/xbmc/xbmc/pull/13034

Reply:
Nice to get to know about such an important change this side way. :-/

Reply:
You was linked as reviewer there.

Reply:
Whoever requested my review there seems not to know that I almost know nothing about cmake. ;-)

Reply:
> Right, but why should I call delete if I know it will do nothing?

Right, but it's a superfluous check because "delete" already checks if the ptr is null. Removing it would (in my opinion) make the code more readable. Just my opinion, though :-)
But that's up to the developer.